### PR TITLE
Clarify&Optimize light constraint reduction pass

### PR DESCRIPTION
**Pull request by earhart**

### CODE REVIEW

**Comment by flaub** [comment]
I recommend @tzerrell or @shiyao00 review

File: tile/codegen/cstr_reduction.cc [hunk 1]
```diff
@@ -34,76 +39,169 @@ static void EvaluatePolynomial(const Polynomial<int64_t> orig_poly, const AliasM
...
+      //     coeff*idx + (rest of the constraint) >= 0
+      // And assuming:
+      //     0 < coeff
+      // .. i.e. when the constraint is == min_value, idx == 0
+      //    i.e. increasing idx only makes the overall constraint increase.
+      // Then: coeff*idx >= -rest
+      //       idx >= -rest / coeff  (rounding division upwards -- N.B. coeff is positive)
+      //       idx >= RoundUp(-rest, coeff)
+      // And since the minimum value (rest) might take on is min_value,
+      //       idx >= RoundUp(-min_value, coeff)
```
Comment (tzerrell) [inline tile/codegen/cstr_reduction.cc]: Something's wrong here but I don't know why. Will get back to you later.

File: tile/codegen/cstr_reduction.cc [hunk 2]
```diff
@@ -15,7 +20,7 @@ using namespace math;    // NOLINT
 using namespace bilp;    // NOLINT
 
 static void EvaluatePolynomial(const Polynomial<int64_t> orig_poly, const AliasMap& alias_map, int64_t* min_value,
-                               int64_t* max_value) {
+                               int64_t* lim_value) {
   Polynomial<int64_t> poly = orig_poly.sym_eval(alias_map.idx_sources());
   int64_t min = poly.constant();
```
[the remark this answers is not in the document]
Comment (earhart) [reply tile/codegen/cstr_reduction.cc]: Done

File: tile/codegen/cstr_reduction.cc [hunk 3]
```diff
@@ -34,76 +39,169 @@ static void EvaluatePolynomial(const Polynomial<int64_t> orig_poly, const AliasM
...
+      // And assuming:
+      //     coeff < 0
+      // .. i.e. when the constraint is == lim_value-1, idx == 0
+      //    i.e. increasing idx only makes the overall constraint decrease.
+      // Then: coeff*idx >= -rest
+      //       -coeff*idx <= rest
+      //       idx <= rest / -coeff  (rounding division upwards -- N.B. -coeff is positive)
+      //       idx <= RoundUp(rest, -coeff)
+      // And since the maximum value (rest) might take on is lim_value-1,
+      //       idx <= RoundUp(lim_value-1, -coeff)
```
[the remark this answers is not in the document]
Comment (earhart) [reply tile/codegen/cstr_reduction.cc]: Done

File: tile/codegen/cstr_reduction.cc [hunk 4]
```diff
@@ -34,76 +39,169 @@ static void EvaluatePolynomial(const Polynomial<int64_t> orig_poly, const AliasM
...
+      // Check if we can reduce the index range by considering the constraint's min value.
+      //
+      // Given an idx:
+      //     coeff*idx + (rest of the constraint) >= 0
+      // And assuming:
+      //     0 < coeff
+      // .. i.e. when the constraint is == min_value, idx == 0
+      //    i.e. increasing idx only makes the overall constraint increase.
+      // Then: coeff*idx >= -rest
+      //       idx >= -rest / coeff  (rounding division upwards -- N.B. coeff is positive)
```
[the remark this answers is not in the document]
Comment (earhart) [reply tile/codegen/cstr_reduction.cc]: Done

File: tile/codegen/cstr_reduction.cc (hunk 1 above)
Comment (earhart) [reply tile/codegen/cstr_reduction.cc]: Fixed, I think.  It was actually a bit trickier to work through than what we were thinking, although the final equation is actually pretty clear (IMHO).

File: tile/codegen/cstr_reduction.cc [hunk 5]
```diff
@@ -34,76 +39,169 @@ static void EvaluatePolynomial(const Polynomial<int64_t> orig_poly, const AliasM
...
+  // We use a forward iterator for this to make the iterator updates upon erasure more explicit.
+
+  // TODO: It would be better to update the original alias_map's index ranges, so that inner blocks would
+  // observe the constrained index space.
+  AliasMap adjusted_alias_map{*alias_map.parent_alias_map(), block};
+  for (auto c_it = block->constraints.end(); c_it != block->constraints.begin();) {
+    --c_it;
+    int64_t min_value, lim_value;
+    EvaluatePolynomial(*c_it, adjusted_alias_map, &min_value, &lim_value);
+    if (0 <= min_value && 0 < lim_value) {
```
[the remark this answers is not in the document]
Comment (earhart) [reply tile/codegen/cstr_reduction.cc]: Done

File: tile/codegen/cstr_reduction.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (earhart) [reply tile/codegen/cstr_reduction.cc]: Done

File: tile/codegen/cstr_reduction.cc [hunk 6]
```diff
@@ -15,98 +21,206 @@ using namespace math;    // NOLINT
...
+                sidx.affine.substitute(idx_name, new_idx);
+              }
+            } else if (auto li = LoadIndex::Downcast(stmt)) {
+              li->from.substitute(idx_name, new_idx);
+            }
+          }
+          min_value += (shift * coeff);
+          lim_value += (shift * coeff);
+          IVLOG(4, "  New idx=" << *idx);
+          IVLOG(4, "  New con=" << constraint << " >= 0: min=" << min_value << " lim=" << lim_value);
```
Comment (tzerrell) [inline tile/codegen/cstr_reduction.cc]: Why is this the new constraint? I don't think you actually changed the value of `constraint` in this section.

File: tile/codegen/cstr_reduction.cc [hunk 7]
```diff
@@ -15,98 +21,206 @@ using namespace math;    // NOLINT
...
+            if (auto sub = Block::Downcast(stmt)) {
+              for (auto& sidx : block->idxs) {
+                sidx.affine.substitute(idx_name, new_idx);
+              }
+            } else if (auto li = LoadIndex::Downcast(stmt)) {
+              li->from.substitute(idx_name, new_idx);
+            }
+          }
+          min_value += (shift * coeff);
+          lim_value += (shift * coeff);
```
Comment (tzerrell) [inline tile/codegen/cstr_reduction.cc]: I don't think that performing a shift changes `lim_value`.

File: tile/codegen/cstr_reduction.cc [hunk 8]
```diff
@@ -15,98 +21,206 @@ using namespace math;    // NOLINT
...
-      if (idx->has_tags(skip_idxs)) {
+      if (idx->has_any_tags(skip_idxs)) {
+        IVLOG(4, "  Skipping index=" << *idx << " (skip_idxs=" << skip_idxs << ")");
         continue;
       }
-      int64_t coeff = -kvp.second;
-      if (idx->affine.getMap().empty() && coeff > 0) {
+      if (!idx->affine.getMap().empty()) {
+        IVLOG(4, "  Skipping non-constant index=" << *idx);
+      }
```
[the remark this answers is not in the document]
Comment (earhart) [reply tile/codegen/cstr_reduction.cc]: Good catch -- I missed this in while refactoring.

File: tile/codegen/cstr_reduction.cc (hunk 6 above)
Comment (earhart) [reply tile/codegen/cstr_reduction.cc]: Substituting `new_idx` for `idx_name` in all of the block's constraints will have updated the current constraint.

File: tile/codegen/cstr_reduction.cc (hunk 7 above)
Comment (earhart) [reply tile/codegen/cstr_reduction.cc]: I think you're right -- subtracting the shift from the index's range and then adding it to the uses of the index should cancel out changes to the max and lim values that can be reached by the constraint equation.  Done.

File: tile/codegen/cstr_reduction.cc [hunk 9]
```diff
@@ -15,98 +21,206 @@ using namespace math;    // NOLINT
...
+      //
+      // Note that rewriting the use of the index within the current constraint will also affect the current
+      // min_value and lim_value, by adding the shift to each.
+      //
+      // Also, note that reducing the range of the index may cause other constraints (potentially
+      // already-processed contraints) to become redundant -- which is why we perform the redundant constraint
+      // check after squeezing indicies across all constraints.
+      //
+      // TODO: Consider removing the index entirely when when shift == range(idx), since the index will only
+      //       ever take on the value of zero.
```
[the remark this answers is not in the document]
Comment (earhart) [reply tile/codegen/cstr_reduction.cc]: Done

**Pull request merged**